packages/plugins/other/visitor-plugin-common/src/utils.ts

```diff
@@ -297,28 +314,87 @@ export function buildScalars(
       .map((scalarType: GraphQLScalarType) => {
         const { name } = scalarType;
         if (typeof scalarsMapping === 'string') {
-          const value = parseMapper(scalarsMapping + '#' + name, name);
-          result[name] = value;
+          const inputMapper = parseMapper(scalarsMapping + '#' + name, name);
+
+          if (inputMapper.isExternal) {
+            inputMapper.type += "['input']";
+          }
+
+          const outputMapper = parseMapper(scalarsMapping + '#' + name, name);
+          if (outputMapper.isExternal) {
+            outputMapper.type += "['output']";
+          }
+
+          result[name] = {
+            input: inputMapper,
+            output: outputMapper,
+          };
         } else if (scalarsMapping && typeof scalarsMapping[name] === 'string') {
-          const value = parseMapper(scalarsMapping[name], name);
-          result[name] = value;
-        } else if (scalarsMapping?.[name]) {
+          const normalizedScalar = normalizeScalarType(scalarsMapping[name]);
+
+          const inputMapper = parseMapper(normalizedScalar.input, name);
+          if (inputMapper.isExternal) {
+            inputMapper.type += "['input']";
+          }
+
+          const outputMapper = parseMapper(normalizedScalar.output, name);
+          if (outputMapper.isExternal) {
+            outputMapper.type += "['output']";
+          }
+
           result[name] = {
-            isExternal: false,
-            type: JSON.stringify(scalarsMapping[name]),
+            input: inputMapper,
+            output: outputMapper,
           };
+        } else if (scalarsMapping?.[name]) {
+          const mappedScalar = scalarsMapping[name];
+          if (typeof mappedScalar === 'object' && mappedScalar.input && mappedScalar.output) {
+            result[name] = {
+              input: {
+                isExternal: false,
+                type: mappedScalar.input,
+              },
+              output: {
+                isExternal: false,
+                type: mappedScalar.output,
+              },
+            };
+          } else {
+            result[name] = {
+              input: {
+                isExternal: false,
+                type: JSON.stringify(mappedScalar),
+              },
+              output: {
+                isExternal: false,
+                type: JSON.stringify(mappedScalar),
+              },
+            };
```

Comment:
Previously, this was stringifying any object:

```ts
result[name] = {
  isExternal: false,
  type: JSON.stringify(scalarsMapping[name]),
};
```

I cannot find tests related to this case. However, this behaviour is backward compatible if the mapper object does not have input/output fields.